File: Algorithms/homework3/src/Main.java
import java.io.*;
import java.util.Scanner;

public class Main {
    static int[] arr = new int[1000];
    static int cache;
    static ObjectOutputStream out;
    static ObjectInputStream in;
    static Scanner sc;
    public static void main(String[] args) throws IOException, ClassNotFoundException {
        //создание массива с рандомными значеиями от 1 до 1000
  /*      for (int i = 0; i <arr.length ; i++) {
            arr[i] = (int)Math.round(Math.random()*1000);
        }
        out = new ObjectOutputStream(new FileOutputStream("myArr.txt"));
        out.writeObject(arr);
        out.close()*/
        in = new ObjectInputStream(new FileInputStream("myArr.txt"));
        arr = (int[])in.readObject();
        in.close();

        printArray(arr);
        System.out.println();
        System.out.println("число операций при пузырьковой сортировке" + boubleSort(arr));//1236193
        System.out.println();
        printArray(arr);

        in = new ObjectInputStream(new FileInputStream("myArr.txt"));
        arr = (int[])in.readObject();
        in.close();

        System.out.println();
        System.out.println("число операций при улучшеной пузырьковой сортировке" + boubleSortMod(arr));//736693
        System.out.println();
        printArray(arr);

        in = new ObjectInputStream(new FileInputStream("myArr.txt"));
        arr = (int[])in.readObject();
        in.close();

        System.out.println();
        System.out.println("число операций при шейкерной сортировке" + shakerSort(arr));//1236193
        printArray(arr);
        System.out.println();

        System.out.println("Введите число для поиска");
        sc = new Scanner(System.in);
        int val = sc.nextInt();
        System.out.println("Индекс " + val + " в массиве " + binarySearch(arr, val));
    }

    static void printArray(int[] arr){
        for (int i = 0; i <arr.length; i++) {
            System.out.print(arr[i] + " ");
            if(i%40==0)
                System.out.println();
        }
    }

    static int boubleSort(int[] arr){
        int count = 0;
        for (int i = 0; i <arr.length ; i++) {
            for (int j = 0; j <arr.length-1 ; j++) {
                count++;
                if(arr[j]>arr[j+1]){
                    count++;
                    cache = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = cache;
                }
            }
        }
        return count;
    }

    static int boubleSortMod(int[] arr){
        int count = 0;
        int a = 1;
        for (int i = 0; i <arr.length ; i++) {
            for (int j = 0; j <arr.length-a ; j++) {
                count++;
                if(arr[j]>arr[j+1]){
                    count++;
                    cache = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = cache;
                }
            }
            a++;
        }
        return count;
    }

    static int shakerSort(int[] arr){
        int count = 0;
        int a = 1;
        for (int i = 0; i <arr.length ; i++) {
            for (int j = 0; j <arr.length-a; j++) {
                count++;
                if(arr[j]>arr[j+1]){
                    count++;
                    cache = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = cache;
                }
            }
            for (int j = arr.length-a; j >0  ; j--) {
                count++;
                if(arr[j]<arr[j-1]){
                    count++;
                    cache = arr[j];
                    arr[j] = arr[j-1];
                    arr[j-1] = cache;
                }
            }
            a++;
        }
        return count;
    }
    static int binarySearch(int[] arr, int val){
        int L = 0, R=arr.length-1;
        int m = L + (R-L)/2;
        while(L<=R && arr[m] != val){
            if(arr[m]<val)
                L = m + 1;
            else
                R = m - 1;
            m = L + (R-L)/2;

        }
        if(arr[m] == val)
            return m;
        else
        return -1;
    }
}

File: Algorithms/homework3/main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define maxN 1000
void printArray(int arr[]);
int boubleSort(int arr[]);
int boubleSortMod(int arr[]);
int shakerSort(int arr[]);
int binarySearch(int arr[], int vol);
// Зарицкий Антон
int main() {

    srand(time(NULL));
    int arr[maxN];
    FILE *in, *out;
   // in = fopen("data.txt", "w");
    out = fopen("data.txt", "r");
  //  for (int i = 0; i <maxN ; ++i) {
   //     arr[i] = rand()%1000;
     //   fprintf(in,"%d\n",arr[i]);
   // }
    for (int i = 0; i <maxN ; ++i) {
        if(fscanf(out, "%d", &arr[i]) == EOF) break;
    }
    fclose(out);
    puts("array before sort");
    printArray(arr);
    puts("");
    // вопрос, правильно ли сделана шекерная сортировка, почему колличество операций при шекерной и пузырьковой сортьировкой одинаково
    //задание №1
    //printf("count of operations on boubleSort = %d ", boubleSort(arr) );// 1248894
    //printf("count of operations on boubleSortMod = %d ", boubleSortMod(arr) );//749394
    //задание №2
    printf("count of operations on shakerSort = %d ", shakerSort(arr) );//1248894
    puts("");
    printArray(arr);
    //задание №3
    puts("Enter number fo search");
    int val;
    scanf("%d", &val);
    if(binarySearch(arr, val)==-1)
        puts("number is not found");
    else printf("index of %d == %d", val, binarySearch(arr, val));
    return 0;
}
void swap (int *x, int *y){
    int z = *x;
    *x = *y;
    *y = z;
}
void printArray(int arr[]){
   // int size = sizeof(arr)/sizeof(int);
    for(int i = 0; i<maxN; i++){
        printf("%d ", arr[i]);
        if(i%30==0)
            printf("\n");
    }
}
int boubleSort(int arr[]){
    int count = 0;
   // int size = sizeof(arr)/sizeof(int);
    for (int i = 0; i<maxN; ++i) {
        for (int j = 0; j <maxN - 1 ; ++j) {
            count++;
            if(arr[j] > arr[j + 1]){
                count++;
                swap(&arr[j], &arr[j+1]);
            }
        }
    }
    return count;
}

int boubleSortMod(int arr[]){
    int count = 0;
    int a = 1;
    int cache;
    for (int i = 0; i <maxN ; i++) {
        for (int j = 0; j <maxN-a ; j++) {
            count++;
            if(arr[j]>arr[j+1]){
                count++;
                swap(&arr[j], &arr[j+1]);
            }
        }
        a++;
    }
    return count;
}

int shakerSort(int arr[]){
    int count = 0;
    int a = 1;
    for (int i = 0; i <maxN ; i++) {
        for (int j = 0; j <maxN-a; j++) {
            count++;
            if(arr[j]>arr[j+1]){
                count++;
                swap(&arr[j], &arr[j+1]);

            }
        }
        for (int j = maxN-a; j >0  ; j--) {
            count++;
            if(arr[j]<arr[j-1]){
                count++;
                swap(&arr[j], &arr[j-1]);
            }
        }
        a++;
    }
    return count;
}
int binarySearch(int arr[], int val){
    int L = 0, R=maxN-1;
    int m = L + (R-L)/2;
    while(L<=R && arr[m] != val){
        if(arr[m]<val)
            L = m + 1;
        else
            R = m - 1;
        m = L + (R-L)/2;

    }
    if(arr[m] == val)
        return m;
    else
        return -1;
}



